README.md
# GpsdoPi
A Raspberry Pi powered GPSDO

Config
------
Modify include/common.h to control the program behavior.



CHANGE_LOG
V0.0.6
------
2015.02.27
Implementation:
1. counter test - SPI library is working

Enhancement:
1. counter nRST pin control
2. freq checker - filter (wrong) freq readings
3. skipFlag - skip freq reading right after operation (steer, calibration..)

V0.0.5
------
2015.02.25
Implementation:
1. AD7249 test - AD7249 library is working

Bugfix:
1. AD7249 library endian problem

V0.0.4
------
2015.02.21
Implementaion:
1. calibration
2. steer

V0.0.3
------
2015.02.21
Implementation:
1. AD7249 library

Documentation:
1. INSTALL
2. a little more detailed README.md

V0.0.2
------
2015.02.20
Implementation:
1. SPI library

V0.0.1
------
2015.02.19
first commit


test/ad7249_test.c
#include "../include/common.h"
#include "../include/ad7249.h"

int main(void)
{
  uint16_t i = 0;
  struct timespec time1, time2;
  time1.tv_sec = 0;
  time1.tv_nsec = 5000;

  wiringPiSetup();

  dacSetup();
  /*
  for (i=0; i<4096; i++)
  {
    dacWriteBinary(1, i);
    nanosleep(&time1, &time2);
  }
  */
  dacWriteBinary(1, 200);
  // dacClose();

  return 0;
}


test/counter_test.c
#include "../include/common.h"
#include <errno.h>
#include "../include/spi.h"

#define T(x) (((x&0xff)<<24)|((x&0xff00)<<8)|((x&0xff0000)>>8)|((x&0xff000000)>>24))

unsigned char* txBuf;
unsigned char* rxBuf;

uint32_t convertFreq(unsigned char* rxBuf);
bool checkFreq(uint32_t freq);

void isr(void)
{
  int i;
  printf("\nisr\n");
  digitalWrite(COUNTER_CS, LOW);
  spiTransfer(txBuf, rxBuf, COUNTER_LEN);
  digitalWrite(COUNTER_CS, HIGH);

  uint32_t freq = convertFreq(rxBuf);
  if (checkFreq(freq))
  // for (i=0; i<4; i++)
    printf("%u", freq);
}

int main(void)
{
  txBuf = (unsigned char*)malloc(COUNTER_LEN);
  rxBuf = (unsigned char*)malloc(COUNTER_LEN);

  if (wiringPiSetup() < 0)
  {
    fprintf(stderr, "Failed to setup wiringPi: %s\n", strerror(errno));
    return 1;
  }
  pinMode(COUNTER_nRST, OUTPUT);
  
  if (spiSetup(SPI_DEV, SPI_MODE, 8, SPI_FREQ) < 0)
  {
    fprintf(stderr, "Failed to setup SPI: %s\n", strerror(errno));
    return 1;
  }

  digitalWrite(COUNTER_nRST, HIGH);
  digitalWrite(COUNTER_nRST, LOW);
  digitalWrite(COUNTER_nRST, HIGH);
  
  if (wiringPiISR(COUNTER_READY, INT_EDGE_RISING, &isr) < 0)
  {
    fprintf(stderr, "Failed to setup ISR: %s\n", strerror(errno));
    return 1;
  }

  while (1)
  {
    // isr();
    // sleep(2);
  }

  return 0;
}

uint32_t convertFreq(unsigned char* rxBuf)
{
  uint32_t freq;

  memcpy(&freq, rxBuf, COUNTER_LEN);
  freq = T(freq);

  return freq;
}

bool checkFreq(uint32_t freq)
{
  if (FREQ * (1 + CHECK_RANGE) < freq || FREQ * (1 - CHECK_RANGE) > freq)
    return false;
  else
    return true;
}


test/build_ad_test.sh
g++ -o ad7249_test ad7249_test.c ../src/ad7249.c -lwiringPi


src/spi.c
#include "../include/common.h"
#include "../include/spi.h"

int spi;

uint8_t spiMode;
uint8_t spiBits = 0;
uint32_t spiSpeed = 5000000;
uint16_t spiDelay;

int spiSetup(char* spiDevice, uint8_t mode, uint8_t bits, uint32_t speed)
{
  int ret = 0;

  // setup GPIO
  pinMode(COUNTER_CS, OUTPUT);
  
  if (spiDevice != NULL)
    spiDevice = spiDevice;
    
  spiMode = mode;
  spiBits = bits;
  spiSpeed = speed;

  spi = open(spiDevice, O_RDWR);
  if (spi < 0)
    spiAbort("Cannot open SPI device");

  // mode
  ret = ioctl(spi, SPI_IOC_WR_MODE, &spiMode);
  if (ret == -1)
    spiAbort("Cannot set SPI mode");

  ret = ioctl(spi, SPI_IOC_RD_MODE, &spiMode);
  if (ret == -1)
    spiAbort("Cannot get SPI mode");

  // bits
  ret = ioctl(spi, SPI_IOC_WR_BITS_PER_WORD, &spiBits);
  if (ret == -1)
    spiAbort("Cannot set bits per word");

  ret = ioctl(spi, SPI_IOC_RD_BITS_PER_WORD, &spiBits);
  if (ret == -1)
    spiAbort("Cannot get bits per word");

  // speed
  ret = ioctl(spi, SPI_IOC_WR_MAX_SPEED_HZ, &spiSpeed);
  if (ret == -1)
    spiAbort("Cannot set max speed hz");

  ret = ioctl(spi, SPI_IOC_RD_MAX_SPEED_HZ, &spiSpeed);
  if (ret == -1)
    spiAbort("Cannot get max speed hz");

  return ret;
}

int spiTransfer(unsigned char* txBuf, unsigned char* rxBuf, int size)
{
  int ret;

  // SPI is full-duplex
  struct spi_ioc_transfer spiTrans;
  spiTrans.tx_buf = (unsigned long)txBuf;
  spiTrans.rx_buf = (unsigned long)rxBuf;
  spiTrans.len = size;
  spiTrans.delay_usecs = spiDelay;
  spiTrans.speed_hz = spiSpeed;
  spiTrans.bits_per_word = spiBits;

  ret = ioctl(spi, SPI_IOC_MESSAGE(1), &spiTrans);
  if (ret < 1)
    spiAbort("Cannot send SPI message");

  return ret;
}

void spiClose()
{
  close(spi);
}

void spiAbort(const char* errStr)
{
  perror(errStr);
  abort();
}



src/ad7249.c
#include "../include/common.h"
#include "../include/ad7249.h"

int dacSetup()
{
  struct timespec holdTime1, holdTime2;
  holdTime1.tv_sec = 0;
  holdTime1.tv_nsec = 50;
  
  // setup GPIO
  pinMode(AD_SCLK, OUTPUT);
  pinMode(AD_SDIN, OUTPUT);
  pinMode(AD_nSYNC, OUTPUT);

  // init state
  digitalWrite(AD_nSYNC, HIGH);
  digitalWrite(AD_SCLK, HIGH);
  digitalWrite(AD_SDIN, LOW);

  // clear both DACs
  dacWriteBinary(1, (uint16_t)0);
  dacWriteBinary(2, (uint16_t)0);

  if (nanosleep(&holdTime1, &holdTime2) < 0)
    return -1;

  return 1;
}

int dacClose()
{
  dacWriteBinary(1, (uint16_t)0);
  dacWriteBinary(1, (uint16_t)0);

  return 1;
}

int dacWriteBinary(int dacNo, uint16_t binary)
{
  int i, j;
  struct timespec cycleTime1, cycleTime2;
  struct timespec holdTime1, holdTime2;

  cycleTime1.tv_sec = 0;
  cycleTime1.tv_nsec = halfCycle;
  holdTime1.tv_sec = 0;
  holdTime1.tv_nsec = 50;

  // select dac
  if (dacNo == 1)
    binary &= ~(1 << DAC_BITS);
  else // dacNo == 2
    binary |= 40970;
    // we don't use dac 2 for now, so use this as a test
    // you should see 1010000000001010 on your logic analyzer

  // 1. set nSYNC
  digitalWrite(AD_nSYNC, LOW);
  if (nanosleep(&holdTime1, &holdTime2) < 0)
    return -1;

  // 2. write 16-bit word
  for (j=0; j<sizeof(uint16_t) * 8; j++)
  {
    i = sizeof(uint16_t) * 8 - j - 1;

    digitalWrite(AD_SCLK, HIGH);
    
    // setup SDIN
    // MSB first - x86 and ARM CPUs are usually little endian
    // attention should be paid, if we port this program to Arduino
    if (binary & (1 << i))
      digitalWrite(AD_SDIN, HIGH);
    else // 0
      digitalWrite(AD_SDIN, LOW);
    if (nanosleep(&cycleTime1, &cycleTime2) < 0)
      return -1;

    digitalWrite(AD_SCLK, LOW);
    if (nanosleep(&cycleTime1, &cycleTime2) < 0)
      return -1;
  }

  // 3. unset nSYNC
  digitalWrite(AD_nSYNC, HIGH);
  digitalWrite(AD_SCLK, HIGH);
  if (nanosleep(&holdTime1, &holdTime2) < 0)
    return -1;

  return 1;
}


src/gpsdoController.c
#include "../include/common.h"
#include <errno.h>
#include "../include/ad7249.h"
#include "../include/spi.h"

bool exitFlag = false;
bool steerFlag = false;
bool skipFlag = false;

unsigned char* txBuf;
unsigned char* rxBuf;

// calibration data
uint32_t maxCount = 0;
uint32_t minCount = 0;
int calibrateTick = 0;

// steer status
int phrase = 0;
int phraseCount = 0;
int tickCount = 0;
uint64_t freqCount = 0;

uint32_t convertFreq(unsigned char* rxBuf);
void calibrate(uint32_t count);
void steer(uint32_t freq);

void isr(void)
{
  // skip the reading right after adjustment (steer, calibration..)
  if (skipFlag)
  {
    skipFlag = false;
    return;
  }

  // read 4 bytes from SPI
  digitalWrite(COUNTER_CS, LOW);
  spiTransfer(txBuf, rxBuf, COUNTER_LEN);
  digitalWrite(COUNTER_CS, HIGH);

  // compute freq
  uint32_t freq = convertFreq(rxBuf);
  if (!checkFreq(freq))
    return;

  if (steerFlag)
  {
    // steer OCXO
    steer(freq);
  }
  else
  {
    // calibration - get steer range
    calibrate(freq);
  }
}

int main(void)
{
  txBuf = (unsigned char*)malloc(COUNTER_LEN);
  rxBuf = (unsigned char*)malloc(COUNTER_LEN);

  // GPIO setup
  if (wiringPiSetup() < 0)
  {
    fprintf(stderr, "Failed to setup wiringPi: %s\n", strerror(errno));
    return 1;
  }
  pinMode(STABLE_PIN, OUTPUT);
  pinMode(COUNTER_nRST, OUTPUT);
  digitalWrite(COUNTER_nRST, HIGH);
  digitalWrite(COUNTER_nRST, LOW);
  digitalWrite(COUNTER_nRST, HIGH);

  // SPI setup
  // FreqCounter SCLK Fmax > 100MHz
  if (spiSetup(SPI_DEV, SPI_MODE, 8, SPI_FREQ) < 0)
  {
    fprintf(stderr, "Failed to setup SPI: %s\n", strerror(errno));
    return 1;
  }

  // AD7249 setup
  if (dacSetup() < 0)
  {
    fprintf(stderr, "Failed to setup DAC\n");
    return 1;
  }

  // finally, start counter interrupt
  if (wiringPiISR(COUNTER_READY, INT_EDGE_RISING, &isr) < 0)
  {
    fprintf(stderr, "Failed to setup ISR: %s\n", strerror(errno));
    return 1;
  }

  while(!exitFlag)
  {
    // command system
    // however, this program could never exit
  }

  // release resources
  free(txBuf);
  free(rxBuf);
  spiClose();
  dacClose();

  return 0;
}

uint32_t convertFreq(unsigned char* rxBuf)
{
  uint32_t freq;
  memcpy(&freq, rxBuf, COUNTER_LEN);
  REVERSE(freq);

  return freq;
}

void calibrate(uint32_t count)
{
  if ((calibrateTick > 0) && (calibrateTick <= 5))
  {
    minCount += count;
    if (calibrateTick == 5)
    {
      dacWriteBinary(AD_DAC, (uint16_t)4095);
      minCount /= 5;

      skipFlag = true;
    }
  }
  else if (calibrateTick < 11)
  {
    maxCount += count;
    if (calibrateTick == 10)
    {
      maxCount /= 5;
      steerFlag = true;

      skipFlag = true;
    }
  }

  calibrateTick++;
}

void steer(uint32_t freq)
{
  uint64_t diff = 0;
  uint64_t min = 0;
  uint64_t max = 0;
  double ratio = 0.0;
  uint16_t correction = 0;

  int count = SteerCounts[STEER_PROCESS][phrase];

  if (tickCount < pow(10, phrase))
  {
    freqCount += freq;
    tickCount++;
  }
  else
  {
    // steer
    skipFlag = true;

    diff = FREQ * (tickCount + 1) - freqCount;
    min = minCount * (tickCount + 1);
    max = maxCount * (tickCount + 1);
    if ((freqCount < minCount) || (freqCount > maxCount))
    {
      fprintf(stderr, "%dHz Out of steer range [%d - %d]Hz\n", freqCount, min, max);
      return;
    }

    ratio = diff / (max - min);
    correction = (uint16_t)(ratio * 4096);
    dacWriteBinary(AD_DAC, correction);

    phraseCount++;
    tickCount = 0;
  }

  // stable
  if (SteerStableFlags[STEER_PROCESS][phrase] == phraseCount)
    digitalWrite(STABLE_PIN, STABLE_LEVEL);

  // next phrase
  if ((count != -1) && (phraseCount == count))
  {
    phrase++;
    phraseCount = 0;
  }
}

bool checkFreq(uint32_t freq)
{
  if (FREQ * (1 + CHECK_RANGE) < freq || FREQ * (1 - CHECK_RANGE) > freq)
    return false;
  else
    return true;
}


include/common.h
#ifndef COMMON_H
#define COMMON_H

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/types.h>
#include <linux/spi/spidev.h>
#include <time.h>

#include <wiringPi.h>

// target frequency 
#define FREQ 10000000L
#define CHECK_RANGE 0.05f // +-5%

// counter pins
#define COUNTER_nRST 29
#define COUNTER_CS 27
#define COUNTER_READY 28
#define COUNTER_LEN 4

// SPI configs
#define SPI_FREQ 5000
#define SPI_DEV "/dev/spidev0.0"
#define SPI_MODE SPI_CPHA | SPI_CPOL

// AD7249 pins
#define AD_SCLK 22
#define AD_SDIN 23
#define AD_nSYNC 21

// AD7249 interface configs
// AD7249 t1(min) = 200ns
#define AD_FREQ 5000000
#define AD_DAC 1

// steer configs
/* there are 2 kinds of steer process
   No.1 - fast (archive 10e-8 in 60s, for audiophile uses)
   No.2 - slower but more precise (archive 10e-10 in 27mins, must use double oven OCXO, for instrumental uses)
*/

#define STEER_PROCESS 0
#define STABLE_PIN 4
#define STABLE_LEVEL LOW

const int SteerCounts[2][4] = {
  {10, 5, 6, -1},
  {0, 6, 6, -1}
};

const int SteerStableFlags[2][4] = {
  {-1, 5, -1, -1},
  {-1, -1, -1, 1}
};

#define REVERSE(x) (((x&0xff)<<24)|((x&0xff00)<<8)|((x&0xff0000)>>8)|((x&0xff000000)>>24))

#endif
